### PR TITLE
Check runtime id length and add hash table comment.

### PR DESCRIPTION
**Pull request by zhangh43**

### CODE REVIEW

File: tests/test_wrong_config.sh [hunk 1]
```diff
@@ -285,6 +285,22 @@ EOF
...
+  echo "Test long runtime id"
+  cat >/tmp/bad_xml_file << EOF
+<?xml version="1.0" ?>
+<configuration>
+    <runtime>
+        <id>plc_python_shared_toooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo_long</id>
+        <image>pivotaldata/plcontainer_python:0.1</image>
+        <command>./client</command>
+        <shared_directory access="ro" container="/clientdir" host="/home/gpadmin/gpdb.devel/bin/plcontainer_clients"/>
+        <setting use_container_logging="yes"/>
```
Comment (paul-guo-) [inline tests/test_wrong_config.sh]: use_container_logging and shared_directory is really not used here for testing.

File: tests/test_wrong_config.sh [hunk 2]
```diff
@@ -285,6 +285,22 @@ EOF
 }
 
 f18 () {
+  echo "Test long runtime id"
```
Comment (paul-guo-) [inline tests/test_wrong_config.sh]: This is a negative case so maybe mention this? e.g. ' runtime id length exceeds allowed size' or others?

File: src/plc_configuration.h [hunk 3]
```diff
@@ -35,6 +35,10 @@ typedef struct plcSharedDir {
 	plcFsAccessMode mode;
 } plcSharedDir;
 
+/*
+ * Struct runtimeConfEntry is the entry of hash table.
+ * the key of hash table must be the fisrt field of struct.
```
Comment (paul-guo-) [inline src/plc_configuration.h]: the -> The
fisrt -> first.

File: src/plc_configuration.h [hunk 4]
```diff
@@ -14,7 +14,7 @@
 #include "plcontainer.h"
 
 #define PLC_PROPERTIES_FILE "plcontainer_configuration.xml"
```
Comment (paul-guo-) [inline src/plc_configuration.h]: I just found plc_refresh_container_config() is still using the hardcoded file name, please fix that.

    sprintf(filename, "%s/plcontainer_configuration.xml", data_directory);

File: src/plc_configuration.c [hunk 5]
```diff
@@ -68,7 +68,13 @@ static void init_runtime_configurations() {
 	hash_ctl.keysize = RUNTIME_ID_MAX_LENGTH;
 	hash_ctl.entrysize = sizeof(runtimeConfEntry);
 	hash_ctl.hash = string_hash;
-
+	/*
+	 * Key and value of hash table share the same storage of entry.
+	 * the first keysize bytes of the entry is the hash key, and the
+	 * value if the entry itself.
+	 * For string key, we use string_hash to caculate hash key and
+	 * use strlcpy to copy key.
```
Comment (paul-guo-) [inline src/plc_configuration.c]: Maybe mention that strlcpy is automatically set as the copy key function due to string_hash?

File: src/plc_configuration.c [hunk 6]
```diff
@@ -130,7 +136,9 @@ static void parse_runtime_configuration(xmlNode *node) {
 		if (id_num == 0) {
 			plc_elog(ERROR, "tag <id> must be specified in configuartion");
 		}
-
+		if ( strlen((char *)runtime_id) >= RUNTIME_ID_MAX_LENGTH) {
+			plc_elog(ERROR, "runtimeid should not be longer than 64 bytes.");
```
Comment (paul-guo-) [inline src/plc_configuration.c]: runtimeid -> runtime id

File: src/containers.c [hunk 7]
```diff
@@ -601,6 +600,10 @@ char *parse_container_meta(const char *source) {
 	 * Allocate container id variable and copy container id 
 	 * the character length of id is last-first.
 	 */
+
+	if (last - first + 1 + 1 >= RUNTIME_ID_MAX_LENGTH) {
+                plc_elog(ERROR, "Runtimeid should not be longer than 64 bytes.");
+        }
```
Comment (paul-guo-) [inline src/containers.c]: No space before }

File: src/containers.c [hunk 8]
```diff
@@ -601,6 +600,10 @@ char *parse_container_meta(const char *source) {
 	 * Allocate container id variable and copy container id 
 	 * the character length of id is last-first.
 	 */
+
+	if (last - first + 1 + 1 >= RUNTIME_ID_MAX_LENGTH) {
+                plc_elog(ERROR, "Runtimeid should not be longer than 64 bytes.");
```
Comment (paul-guo-) [inline src/containers.c]: Runtimeid -> Runtime id

**Comment by paul-guo-** [comment]
The plcontainer utility needs to check this also (both python code and test).

File: src/plc_configuration.c (hunk 6 above)
Comment (HaozhouWang) [inline src/plc_configuration.c]: it is better to change to  
runtime id should not be longer than 63 characters.

File: src/plc_configuration.h (hunk 4 above)
Comment (zhangh43) [reply src/plc_configuration.h]: fixed

File: tests/test_wrong_config.sh (hunk 1 above)
Comment (zhangh43) [reply tests/test_wrong_config.sh]: no harm here

**Comment by zhangh43** [comment]
update the plcontainer script too.

File: management/bin/plcontainer [hunk 9]
```diff
@@ -457,6 +457,10 @@ def validate_xml(filename):
                 raise Exception("Validation failed")
 
             runtime_id = runtime.find('id').text
+
+            if len(runtime_id) >= 64:
+                logger.error("length of runtime id %s should not exceeds 64.", runtime_id)
```
Comment (paul-guo-) [inline management/bin/plcontainer]: exceed.
%s -> '%s'

File: src/plc_configuration.c [hunk 10]
```diff
@@ -426,7 +426,7 @@ static int plc_refresh_container_config(bool verbose) {
 	LIBXML_TEST_VERSION
 
 	/* Parse the file and get the DOM */
-	sprintf(filename, "%s/plcontainer_configuration.xml", data_directory);
+	sprintf(filename, "%s/"#PLC_PROPERTIES_FILE"", data_directory);
```
Comment (paul-guo-) [inline src/plc_configuration.c]: "%s/" PLC_PROPERTIES_FILE
Should be ok.

**Comment by paul-guo-** [comment]
So the runtime id length must <=63 or <=64?

**Comment by zhangh43** [comment]
Maybe 63 is better. I'll change the message on plcontainer script.

**Comment by zhangh43** [comment]
RUNTIME_ID_MAX_LENGTH is 64, and considering \0, the real string length should be <=63

File: src/containers.c [hunk 11]
```diff
@@ -601,8 +601,8 @@ char *parse_container_meta(const char *source) {
 	 * the character length of id is last-first.
 	 */
 
-	if (last - first + 1 + 1 >= RUNTIME_ID_MAX_LENGTH - 1) {
-		plc_elog(ERROR, "Runtime id should not be longer than 64 bytes.");
+	if (last - first + 1 + 1 > RUNTIME_ID_MAX_LENGTH - 1) {
+		plc_elog(ERROR, "Runtime id should not be longer than 63 bytes.");
 	}
```
Comment (paul-guo-) [inline src/containers.c]: Should be

last - first + 1 + 1 > RUNTIME_ID_MAX_LENGTH

**Comment by paul-guo-** [comment]
+1

**Pull request merged**